internal/policy/verify.go

```diff
@@ -329,6 +329,119 @@ func VerifyRelativeForRef(ctx context.Context, repo *gitinterface.Repository, fi
 	return nil
 }
 
+// VerifyMergeable checks if a change in a featureRef can be merged into the
+// targetRef. For Git reference protection rules, if a rule can be met with one
+// additional signature, this function returns true and without errors to
+// indicate that the merge will meet policy _if_ the resultant RSL entry is
+// signed by an approved party.
+func VerifyMergeable(ctx context.Context, repo *gitinterface.Repository, targetRef, featureRef string) (bool, error) {
+	slog.Debug("Loading policy...")
+	policyState, err := LoadCurrentState(ctx, repo, PolicyRef)
+	if err != nil {
+		return false, err
+	}
+
+	slog.Debug("Loading current set of attestations...")
+	attestationsState, err := attestations.LoadCurrentAttestations(repo)
+	if err != nil {
+		return false, err
+	}
+
+	var fromID gitinterface.Hash
+	slog.Debug(fmt.Sprintf("Identifying latest RSL entry for '%s'...", targetRef))
+	targetEntry, _, err := rsl.GetLatestReferenceEntry(repo, rsl.ForReference(targetRef), rsl.IsUnskipped())
+	switch {
+	case err == nil:
+		fromID = targetEntry.TargetID
+	case errors.Is(err, rsl.ErrRSLEntryNotFound):
+		fromID = gitinterface.ZeroHash
+	default:
+		return false, err
+	}
+
+	slog.Debug(fmt.Sprintf("Identifying latest RSL entry for '%s'...", featureRef))
+	featureEntry, _, err := rsl.GetLatestReferenceEntry(repo, rsl.ForReference(featureRef), rsl.IsUnskipped())
+	if err != nil {
+		return false, err
+	}
+
+	mergeTreeID, err := repo.GetMergeTree(fromID, featureEntry.TargetID)
+	if err != nil {
+		return false, err
+	}
+
+	authorizationAttestation, approverKeyIDs, err := getApproverAttestationAndKeyIDsForIndex(ctx, repo, policyState, attestationsState, targetRef, fromID, mergeTreeID)
+	if err != nil {
+		return false, err
+	}
+
+	verifiers, err := policyState.FindVerifiersForPath(fmt.Sprintf("%s:%s", gitReferenceRuleScheme, targetRef))
+	if err != nil {
+		return false, err
+	}
+
+	_, rslEntrySignatureNeededForThreshold, err := verifyGitObjectAndAttestationsUsingVerifiers(ctx, verifiers, gitinterface.ZeroHash, authorizationAttestation, approverKeyIDs, true)
+	if err != nil {
+		return false, fmt.Errorf("not enough approvals to meet Git namespace policies, %w", ErrUnauthorizedSignature)
+	}
+
+	hasFileRule, err := policyState.hasFileRule()
+	if err != nil {
+		return false, err
+	}
+	if !hasFileRule {
+		return rslEntrySignatureNeededForThreshold, nil
```

Comment:
> @wlynch: Also not clicking what this has to do with file rules. If this is for mergeability, e.g. "the commit has the number of required RSL entries needed to meet the threshold", then this should be renamed since it reads like "RSL entries are needed to meet the threshold"

That's actually accurate, we're responding saying it's mergeable _as long as the RSL entry for the merge is by an authorized user_. IMO this flexibility is useful when merges may be outsourced to bots and all approvals come in via attestations. We may not want to add the bot as authorized to the rule since that might be too much power.

Reply:
This is consumed in repository here: https://github.com/gittuf/gittuf/pull/455/files#diff-62c1c7d5c8c9f9d2c194f625d419809db12b1a2b514bc40dd8453a092c2571afR154-R158

Maybe that needs to be slog.Info so it's always printed, or plumbed through to cmd so it can print to stdout. We could also exit with a different exit code for programmatic knowledge. :thinking:

Reply:
I think this has been resolved, we also don't have a cmd to plumb this through to. 😄